ebpf/state.go

```diff
@@ -0,0 +1,209 @@
+package ebpf
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/cihub/seelog"
+)
+
+var _ NetworkState = &networkState{}
+
+const (
+	// DEBUGCLIENT is the ClientID for debugging
+	DEBUGCLIENT          = -1
+	defaultClientExpiry  = 5 * time.Minute
+	defaultCleanInterval = 10 * time.Second
+)
+
+// NetworkState takes care of handling the logic for:
+// - closed connections
+// - sent and received bytes per connection
+type NetworkState interface {
+	Clients() []int
+	Connections(clientID int) []ConnectionStats
+	StoreConnections(conns []ConnectionStats)
+	StoreClosedConnection(conn ConnectionStats)
+	RemoveClient(clientID int) error
+}
+
+type sendRecvStats struct {
+	totalSent uint64
+	lastSent  uint64
+
+	totalRecv uint64
+	lastRecv  uint64
+
+	totalRetransmits uint32
+	lastRetransmits  uint32
+}
+
+type client struct {
+	id                int
+	lastFetch         time.Time
+	closedConnections []*ConnectionStats
+	stats             map[string]*sendRecvStats
+}
+
+type networkState struct {
+	clients      map[int]*client
+	clientsMutex sync.Mutex
+
+	connections []ConnectionStats
+	connsMutex  sync.Mutex
+
+	cleanInterval time.Duration
+	clientExpiry  time.Duration
+}
+
+// NewDefaultNetworkState creates a new network state with default settings
+func NewDefaultNetworkState() NetworkState {
+	return NewNetworkState(defaultCleanInterval, defaultClientExpiry)
+}
+
+// NewNetworkState creates a new network state
+func NewNetworkState(cleanInterval time.Duration, clientExpiry time.Duration) NetworkState {
+	ns := &networkState{
+		clients:       map[int]*client{},
+		connections:   []ConnectionStats{},
+		cleanInterval: cleanInterval,
+		clientExpiry:  clientExpiry,
+	}
+
+	// Start tracking expiry time for clients
+	go func() {
+		for now := range time.NewTicker(ns.cleanInterval).C {
+			ns.trackClientExpiry(now)
+		}
+	}()
+
+	return ns
+}
+
+func (ns *networkState) Clients() []int {
+	ns.clientsMutex.Lock()
+	defer ns.clientsMutex.Unlock()
+	clients := make([]int, 0, len(ns.clients))
+
+	for id := range ns.clients {
+		clients = append(clients, id)
+	}
+
+	return clients
+}
+
+func (ns *networkState) Connections(id int) []ConnectionStats {
+	ns.connsMutex.Lock()
+	defer ns.connsMutex.Unlock()
+
+	if old := ns.newClient(id); !old {
+		// First time we see this client, use global state
+		return ns.connections
+	}
+
+	conns := removeDuplicates(append(ns.closedConns(id), ns.connections...))
```

Comment:
What happens when a connection is closed + reopened on the same ports during the interval? 

What values do we use + send here? The sum of their counts, the new connections counts, or the old connections counts? What should we send?